acme/src/finalize.rs

```diff
@@ -26,37 +30,163 @@ impl RustyAcme {
+        let csr = x509_cert::request::CertReq {
+            info: cert_info,
+            algorithm,
+            signature,
         };
-        params.key_pair = Some(rcgen::KeyPair::from_pem(kp.as_str())?);
-        params.subject_alt_names = vec![
-            rcgen::SanType::URI(identifier.client_id.to_lowercase()),
-            rcgen::SanType::URI(identifier.handle.to_lowercase()),
-        ];
-
-        let cert = rcgen::Certificate::from_params(params)?;
-        let csr = cert.serialize_request_der()?;
+        let csr = csr.to_der()?;
         let csr = base64::prelude::BASE64_URL_SAFE_NO_PAD.encode(csr);
         Ok(csr)
     }
 
+    fn csr_alg(alg: JwsAlgorithm) -> RustyAcmeResult<x509_cert::spki::AlgorithmIdentifierOwned> {
+        let oid = match alg {
+            JwsAlgorithm::Ed25519 => oid_registry::OID_SIG_ED25519,
+            JwsAlgorithm::P256 => oid_registry::OID_SIG_ECDSA_WITH_SHA256,
+            JwsAlgorithm::P384 => oid_registry::OID_SIG_ECDSA_WITH_SHA384,
+        };
+        Self::into_asn1_alg(oid, None)
+    }
+
+    fn csr_subject(identifier: &WireIdentifier) -> RustyAcmeResult<x509_cert::name::DistinguishedName> {
+        let dn_domain_oid = oid_registry::OID_X509_ORGANIZATION_NAME.as_bytes().try_into()?;
+        let dn_domain_value =
+            x509_cert::attr::AttributeValue::new(x509_cert::der::Tag::Utf8String, identifier.domain.as_bytes())?;
+        let dn_domain = x509_cert::attr::AttributeTypeAndValue {
+            oid: dn_domain_oid,
+            value: dn_domain_value,
+        };
+
+        // TODO: temporarily using a custom OIDC for carrying the display name without having it listed as a DNS SAN.
+        // reusing LDAP's OID for diplay name see http://oid-info.com/get/2.16.840.1.113730.3.1.241
+        let dn_display_name_oid = asn1_rs::oid!(2.16.840 .1 .113730 .3 .1 .241).as_bytes().try_into()?;
+        let dn_display_name_value =
+            x509_cert::attr::AttributeValue::new(x509_cert::der::Tag::Utf8String, identifier.display_name.as_bytes())?;
+        let dn_display_name = x509_cert::attr::AttributeTypeAndValue {
+            oid: dn_display_name_oid,
+            value: dn_display_name_value,
+        };
+
+        let domain = x509_cert::name::RelativeDistinguishedName(vec![dn_domain].try_into()?);
+        let display_name = x509_cert::name::RelativeDistinguishedName(vec![dn_display_name].try_into()?);
+        let subject = x509_cert::name::DistinguishedName::from(vec![domain, display_name]);
+        Ok(subject)
+    }
+
+    fn csr_spki(alg: JwsAlgorithm, kp: &Pem) -> RustyAcmeResult<x509_cert::spki::SubjectPublicKeyInfoOwned> {
+        let (pk, algorithm) = match alg {
+            JwsAlgorithm::Ed25519 => {
+                let pk = Ed25519KeyPair::from_pem(kp.as_str())?.public_key().to_bytes();
+                let alg = Self::into_asn1_alg(oid_registry::OID_SIG_ED25519, None)?;
```

Comment:
Can we insert a comment linking/quoting the spec about the format of ed25519/NIST keys that the spec expects? Would be easier to check for correctness (like a part that says EdDSA = ASN.1 and ECDSA = uncompressed raw bytes)